src/sqlfluff/core/plugin/hookspecs.py

```diff
@@ -14,15 +16,24 @@ class PluginSpec:
 
     @hookspec
     @abstractmethod
-    def get_rules(self):
-        """Get plugin rules."""
+    def get_rules(self) -> List[Type[RuleType]]:
+        """Get plugin rules.
+
+        NOTE: While the type annotation for this method returns `Type[RuleType]`
+        all plugin implementations should instead import `sqlfluff.core.rules` and
+        return the more specific `Type[BaseRule]`. This base definition of the
+        PluginSpec cannot import that yet as it would cause a circular import.
+        """
```

Comment:
@WittierDinosaur - this is the only bit that feels a bit gross, but this is the least bad of a few choices. Those options are:
1. (what I've done, which is) Have the plugin module describe something "rule-like" which `get_rules()` returns, and that be a parent class of `BaseRule` so that mypy is happy.
2. We could have the type annotation here reference `BaseRule` directly, but that violates the `lint-imports` rule which says that the `linter` can't depend on `rules` (it must be the other way around only) - and at the moment, `linter` depends on `config` which depends on `plugin`. `linter` is likely always going to depend on `config`, and refactoring `config` so that it doesn't need to depend on `plugin` seems hard or even impossible.
3. We could avoid the new `RuleType` class and just make this type hint really vague (i.e. `List[Any]` or `List[Type[Any]]`). You could argue that the current `RuleType` implementation isn't much better than this because the class doesn't actually define anything other than being a marker for inheritance, but it feels a little more descriptive than this.

What do you reckon?

Reply:
Ignore all of this. I think it's gross. I've just added an exception in the import linter instead.